### ISSUE TITLE
Should BeanShell have automatic type conversions? (was Should BeanShell be type safe?)

### ISSUE THREAD

**Issue by opeongo** [beanshell/beanshell#759]
There are a couple of changes in #721 that remove useful warning messages for the average programmer.  Sure, this PR removes warning from the interpreter, but this might just be hiding programming problems that would be better to be warned about.  For example, auto-converting one data type to another.  This may seem like a minor convenience to save a cast or a perhaps relational binary operator, but my experience in large code bases is that it is nice to have the compiler catching my mistakes... either a typo or referencing the wrong variable.

So if you try to use a number or string as a conditional test in an if statement there should be a warning about this.  If the string to boolean conversion is blank=false/non-blank=true then all this is saving is (a) vs (a.isEmpty()),  And for numbers if it is 0=false/!0=true then the saving is (a) vs (a==0).  Going from boolean to number the saving is (a?1:0).  That is really not many characters to add to make it explicit.  Having the explicit test also improves the clarity of the code, making obvious what would otherwise be something to have to think about.

I have used languages that do these sort of automatic conversions (perl for example).  Most people consider that perl programs are write-only code:  the author can understand the code as it is being written, but wait a few months and the statements make as much sense as modem line noise.  So convenient to write in because so economical of keystrokes.  So terrible to maintain.  This is an important aspect of language design.

So I think that there are two points to consider here:

1. Compiler warnings that indicate that something unexpected is happening often catch unexpected things, and this is a good thing.  This is the benefit of having a type system.  BeanShell is loosely typed which makes it easy to write scripts, and also the type checking messages from the compiler are a nice thing to assist developers. 
2. Clarity is an important aspect of writing program/scripts that are clear and that other people will find easy to read and follow.  Having the compiler perform slight-of-hand automatic conversion reduces that clarity and adds a code smell.  Just like ```((a/b)*c)+d``` is easier to quickly understand than ```a / b * c + d```, unexpected conversions make more sense when they are made explicit.

The automatic type conversion between numbers-string-boolean gives up warnings which is a loss, and in return it really does not add much to the language other than save a minimal number of keystrokes.  The loss of clarity that comes with automatic type conversion really does not help anyone.

I would like to keep the majority of #721 as is, but back out the automatic datatype conversion portions.

**Comment by nickl-** [comment]
> have the compiler catching my mistakes
Is this not an issue that we are not compiling but interpreting? As discussed in #501

Many scripting languages ie. JavaScript is loose typed, with large code bases, without any problems. The convenience outweighs the cons. And you cannot have it both ways, type safe requires type declaration, BeanShell never had type declaration so is therefor by definition loosely typed.

If you want type safe then use JAVA compiled.

Suggestion close issue: works as implemented.

**Comment by nickl-** [comment]
**closed: works as implemented.**

**Comment by opeongo** [comment]
> > have the compiler catching my mistakes
> 
> Is this not an issue that we are not compiling but interpreting?

I think you know what I mean...  that during parse and run time the language system catches and reports obvious errors rather than silently doing something with them.

> JavaScript is loose typed

JavaScript is a very poor example to cite from my friend... [there are so many oddities and irregularities in that language](https://www.reddit.com/r/AskProgramming/comments/lwc3gj/why_a_lot_of_software_engineers_say_javascript_is/).

Let's talk about BeanShell and let me narrow down my comments a bit.  Loose typing is great.   Automatic widening of compatible types is great: float->double, byte->short->int->long no problem.  Even long->BigInteger and double->BigDecimal.

But automatic type conversions from one type species to another (like int->boolean, boolean->int, string->int) seems like a poor choice.  So little gain at the cost of loosing other benefits.

Boolean in Java is truthy concept:  the value can be either true or false.  Interpreting non-booleans as truthy doesn't get much and it opens up the opportunity for uncaught mistakes.  We had boolean evaluation of int's and pointers in C and Perl and Fortran and it turned out that language designers came around to thinking that this wasn't such a good idea after all.  Java wasn't built by dummies.  

So what does it get you?  You can have this
```java
int i = 1;
if (i)
  System.out.println("true");
```

instead of
```java
int i = 1;
if (i!=0)
  System.out.println("true");
```

Big deal.  Language designers talk about features like this 'not pulling their weight'.  It adds cognitive load on developers for very little added benefit.  So you save a couple of keystrokes but end up losing the little bit of type checking that there is.  As far as I am concerned automatic type conversion just makes it easier for bugs to go unnoticed. 

And as far as not pulling their weight, these so-called features also add to BeanShell source code complexity making the code base harder to work on.  Although it might have been easy to implement I wouldn't even call it low hanging fruit, it's more of a net negative.  And the number of hours that I have spent chasing down BeanShell bugs in the past 4 years tells me it is not worth it.

And there are so many other interesting potential BeanShell features that should be higher priority than this.  VarArgs and enum's were great additions to BeanShell that really pulled their weight.  What about new project amber features like pattern matching and records?

**Comment by jimjag** [comment]
Closing during a backlog triage pass: the maintainer explicitly closed this: "works as implemented". Reopen if this is still relevant.

**Comment by opeongo** [comment]
I would like to keep this one open for a bit more discussion.  It was open for discussion, but neither Nick or I had anything more to add so it stalemated.

As I mentioned previously in this issue, the automatic conversion between data types, beyond what java supports, it too weird for a language that is advertised as "Lightweight scripting for Java".  As the slogan says, this scripting language is for Java, it doesn't advertise itself as a general purpose scripting language for the JVM.  

#721 introduced a lot of changes that were not discussed anywhere (unless you consider #501 to be carte blanche to do anything), including the automatic conversion from one data type to another.  I see some of these changes as a regression, removing the distinctions between data types that had been introduced as support of good programming practices.  The Java language architects rejected the C idiom where a word with all zero bits is false otherwise true.  Instead they chose an explicit boolean data type.  If you use an `int` where a `boolean` is expected you get an error message.  Similarly, if you use a `String` where a primitive is expected that is also an error.   The reason for catching type misuse is to detect code smells.  It is cheap and easy to test an `int` for true or false, so the auto conversion feature doesn't real;y add much.  As the Java architects say, some features don't carry their weight, and automatic type conversion is certainly one of them.  These changes have reduced the potential for useful error messages, they hides code smells, and it's almost like
```
javac myfile.java >/dev/null 2>&1
```
Yes, the automatic feature conversion works as implemented, but it adds minimal value to the language at a cost of pointlessly diverging from Java.

I think that the "automatic data type conversion without warning" features should be given more consideration. leading to a rationalized decision about keeping or reverting.

Obviously, my vote is to revert.

**Comment by stefanofornari** [comment]
I agree with @opeongo on type safety; even just 'var' added to Java was not a good idea IMHO. I rather prefer to have an explicit cast with implicit/extensible conversion rather than do everything hidden. I mean that if you write  `A a = (A)b` where `b` is of an incompatible type, the interpreter tries to convert between the types and only if there is no compatible converter it throws an exception.

my 2 cents.
